File: backend/repurpose_render.py
"""
repurpose_render.py — React-to-FFmpeg rendering bridge.

This module is imported by repurpose_router.py to keep the pipeline file clean.
The key function here is render_repurpose_export(), which translates a
RepurposeExportRequest (with normalized Y position) into a concrete FFmpeg
single-pass filter chain at the target 1080×1920 canvas.
"""

import os
import subprocess


def _escape_drawtext(text: str) -> str:
    return (
        str(text or "")
        .replace("\\", "\\\\")
        .replace(":", r"\:")
        .replace("'", r"\'")
        .replace("%", r"\%")
        .replace(",", r"\,")
        .replace("[", r"\[")
        .replace("]", r"\]")
    )


def _resolve_font() -> str:
    windir = os.environ.get("WINDIR", r"C:\Windows")
    local_appdata = os.environ.get("LOCALAPPDATA", "")
    candidates = [
        os.environ.get("SF_FONT_PATH", ""),
        os.path.join(windir, "Fonts", "SF-Pro-Text-Regular.otf"),
        os.path.join(windir, "Fonts", "SF-Pro-Display-Regular.otf"),
        os.path.join(local_appdata, "Microsoft", "Windows", "Fonts", "SF-Pro-Text-Regular.otf"),
        os.path.join(local_appdata, "Microsoft", "Windows", "Fonts", "SF-Pro-Display-Regular.otf"),
        os.path.join(windir, "Fonts", "segoeui.ttf"),
        os.path.join(windir, "Fonts", "arial.ttf"),
        "/System/Library/Fonts/SFNS.ttf",
        "/System/Library/Fonts/SFNSDisplay.ttf",
        "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf",
        "/usr/share/fonts/truetype/liberation2/LiberationSans-Regular.ttf",
    ]
    for p in candidates:
        if p and os.path.exists(p):
            return p
    return ""


def _wrap_lines(text: str, max_non_space: int = 25, max_lines: int = 3) -> list[str]:
    """
    Wrap overlay text: split when non-space chars exceed 32 per line.
    Spaces are never counted toward the limit. Max 3 lines.
    """
    words = str(text or "").replace("\n", " ").split()
    if not words:
        return []
    lines: list[str] = []
    current_words: list[str] = []
    current_len = 0

    for word in words:
        word_len = len(word.replace(" ", ""))
        if current_words and current_len + word_len > max_non_space:
            lines.append(" ".join(current_words))
            if len(lines) >= max_lines:
                break
            current_words = [word]
            current_len = word_len
        else:
            current_words.append(word)
            current_len += word_len

    if current_words and len(lines) < max_lines:
        lines.append(" ".join(current_words))
    return lines


def _run(args: list, step: str) -> dict:
    cmd = ["ffmpeg", "-y"] + args
    try:
        r = subprocess.run(cmd, capture_output=True, text=True, timeout=300)
        if r.returncode != 0:
            return {"success": False, "error": f"FFmpeg [{step}] rc={r.returncode}: {r.stderr[-2000:]}"}
        return {"success": True}
    except subprocess.TimeoutExpired:
        return {"success": False, "error": f"FFmpeg [{step}] timed out"}
    except FileNotFoundError:
        return {"success": False, "error": "ffmpeg not found in PATH"}
    except Exception as e:
        return {"success": False, "error": str(e)}


def render_repurpose_export(
    input_path: str,
    output_path: str,
    background_hex: str,
    overlay_text: str,
    overlay_y_position_normalized: float,
    canvas_w: int = 1080,
    canvas_h: int = 1920,
) -> dict:
    """
    React-to-FFmpeg rendering bridge.

    Parameters match RepurposeExportRequest exactly:
      background_hex                : CSS hex, e.g. '#0d0d1a'
      overlay_text                  : hook string chosen by the user (empty = no text)
      overlay_y_position_normalized : float 0.0 (top) … 1.0 (bottom)

    Single FFmpeg pass:
      [bg] solid color canvas
      [fg] video scaled to fit (letterboxed, preserving AR)
      overlay at center
      drawtext at normalized Y → absolute pixel Y
    """
    # Map normalized Y to pixel space
    norm_y = max(0.02, min(0.98, float(overlay_y_position_normalized)))
    font_size = max(48, int(canvas_h * 0.055))         # ~106 px at 1920h
    line_spacing = max(8, int(canvas_h * 0.012))        # ~23 px at 1920h
    y_px = int(norm_y * canvas_h) - font_size // 2
    y_px = max(10, min(canvas_h - font_size - 10, y_px))

    # ── Step 1: background + letterbox composite ───────────────────────────
    # Normalize hex: strip '#' for FFmpeg color spec
    bg = background_hex.lstrip("#")
    filter_complex = (
        f"color=c=0x{bg}:s={canvas_w}x{canvas_h}:r=30[bg];"
        f"[0:v]scale={canvas_w}:{canvas_h}:force_original_aspect_ratio=decrease,"
        f"setsar=1[fg];"
        f"[bg][fg]overlay=(W-w)/2:(H-h)/2[base]"
    )

    # ── Step 2: drawtext (only if hook text provided) ─────────────────────
    map_label = "[base]"
    if overlay_text and overlay_text.strip():
        font_path = _resolve_font()
        lines = _wrap_lines(overlay_text.strip())
        # Bottom-anchor: block sits 20px from bottom, 24px side margins each side
        font_size    = max(48, int(canvas_h * 0.055))
        line_spacing = max(8, int(canvas_h * 0.012))
        block_height = len(lines) * font_size + max(0, len(lines) - 1) * line_spacing
        base_y       = max(24, canvas_h - block_height - 20)
        current = "base"

        for i, line in enumerate(lines):
            esc      = _escape_drawtext(line)
            line_y   = base_y + i * (font_size + line_spacing)
            next_lbl = f"txt{i}"

            draw = (
                f"[{current}]drawtext="
                f"text='{esc}':"
                f"x=(w-text_w)/2:"
                f"y={line_y}:"
                f"fontsize={font_size}:"
                f"fontcolor=white:"
                f"borderw=5:bordercolor=black@0.92:"
                f"shadowx=3:shadowy=3:shadowcolor=black@0.85"
            )
            if font_path:
                draw += f":fontfile='{_escape_drawtext(font_path)}'"
            draw += f"[{next_lbl}]"

            filter_complex += f";{draw}"
            current = next_lbl

        map_label = f"[{current}]"

    args = [
        "-i", input_path,
        "-filter_complex", filter_complex,
        "-map", map_label,
        "-map", "0:a?",
        "-c:v", "libx264", "-preset", "fast", "-crf", "22",
        "-c:a", "aac", "-b:a", "128k",
        "-pix_fmt", "yuv420p",
        "-movflags", "+faststart",
        output_path,
    ]
    return _run(args, "render_repurpose_export")

File: backend/repurpose_render.js
const fs = require('fs')
const path = require('path')
const { execFile } = require('child_process')

function escapeDrawtext(text) {
  return String(text || '')
    .replace(/\\/g, '\\\\')
    .replace(/:/g, '\\:')
    .replace(/'/g, "\\'")
    .replace(/%/g, '\\%')
    .replace(/,/g, '\\,')
    .replace(/\[/g, '\\[')
    .replace(/\]/g, '\\]')
}

function resolveFont() {
  const windir = process.env.WINDIR || 'C:\\Windows'
  const localAppData = process.env.LOCALAPPDATA || ''
  const candidates = [
    process.env.SF_FONT_PATH || '',
    path.join(__dirname, 'fonts', 'SFProText-Regular.ttf'),
    path.join(windir, 'Fonts', 'SF-Pro-Text-Regular.otf'),
    path.join(windir, 'Fonts', 'SF-Pro-Display-Regular.otf'),
    path.join(localAppData, 'Microsoft', 'Windows', 'Fonts', 'SF-Pro-Text-Regular.otf'),
    path.join(localAppData, 'Microsoft', 'Windows', 'Fonts', 'SF-Pro-Display-Regular.otf'),
    path.join(windir, 'Fonts', 'segoeui.ttf'),
    path.join(windir, 'Fonts', 'arial.ttf'),
    '/System/Library/Fonts/SFNS.ttf',
    '/System/Library/Fonts/SFNSDisplay.ttf',
    '/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf',
    '/usr/share/fonts/truetype/liberation2/LiberationSans-Regular.ttf',
  ]
  return candidates.find(candidate => candidate && fs.existsSync(candidate)) || ''
}

function codePointLength(text) {
  return Array.from(text).length
}

function wrapLines(text, maxNonSpace = 25, maxLines = 3) {
  const words = String(text || '').replace(/\n/g, ' ').trim().split(/\s+/).filter(Boolean)
  if (!words.length) return []
  const lines = []
  let currentWords = []
  let currentLength = 0
  for (const word of words) {
    const wordLength = codePointLength(word.replace(/ /g, ''))
    if (currentWords.length && currentLength + wordLength > maxNonSpace) {
      lines.push(currentWords.join(' '))
      if (lines.length >= maxLines) break
      currentWords = [word]
      currentLength = wordLength
    } else {
      currentWords.push(word)
      currentLength += wordLength
    }
  }
  if (currentWords.length && lines.length < maxLines) lines.push(currentWords.join(' '))
  return lines
}

function run(args, step) {
  return new Promise(resolve => {
    execFile('ffmpeg', ['-y', ...args], { timeout: 300000, maxBuffer: 16 * 1024 * 1024 }, (error, stdout, stderr) => {
      if (!error) {
        resolve({ success: true })
        return
      }
      if (error.killed || error.code === 'ETIMEDOUT') {
        resolve({ success: false, error: `FFmpeg [${step}] timed out` })
        return
      }
      if (error.code === 'ENOENT') {
        resolve({ success: false, error: 'ffmpeg not found in PATH' })
        return
      }
      resolve({ success: false, error: `FFmpeg [${step}] rc=${error.code}: ${String(stderr || '').slice(-2000)}` })
    })
  })
}

async function renderRepurposeExport(
  inputPath,
  outputPath,
  backgroundHex,
  overlayText,
  overlayYPositionNormalized,
  canvasW = 1080,
  canvasH = 1920,
) {
  const normY = Math.max(0.02, Math.min(0.98, Number(overlayYPositionNormalized)))
  let fontSize = Math.max(48, Math.trunc(canvasH * 0.055))
  const lineSpacing = Math.max(8, Math.trunc(canvasH * 0.012))
  let yPx = Math.trunc(normY * canvasH) - Math.trunc(fontSize / 2)
  yPx = Math.max(10, Math.min(canvasH - fontSize - 10, yPx))

  const bg = String(backgroundHex).replace(/^#/, '')
  let filterComplex = (
    `color=c=0x${bg}:s=${canvasW}x${canvasH}:r=30[bg];`
    + `[0:v]scale=${canvasW}:${canvasH}:force_original_aspect_ratio=decrease,`
    + 'setsar=1[fg];'
    + '[bg][fg]overlay=(W-w)/2:(H-h)/2[base]'
  )
  let mapLabel = '[base]'

  if (overlayText && String(overlayText).trim()) {
    const fontPath = resolveFont()
    const lines = wrapLines(String(overlayText).trim())
    fontSize = Math.max(48, Math.trunc(canvasH * 0.055))
    const blockHeight = lines.length * fontSize + Math.max(0, lines.length - 1) * lineSpacing
    const baseY = Math.max(24, canvasH - blockHeight - 20)
    let current = 'base'

    for (let index = 0; index < lines.length; index += 1) {
      const escaped = escapeDrawtext(lines[index])
      const lineY = baseY + index * (fontSize + lineSpacing)
      const nextLabel = `txt${index}`
      let draw = (
        `[${current}]drawtext=`
        + `text='${escaped}':`
        + 'x=(w-text_w)/2:'
        + `y=${lineY}:`
        + `fontsize=${fontSize}:`
        + 'fontcolor=white:'
        + 'borderw=5:bordercolor=black@0.92:'
        + 'shadowx=3:shadowy=3:shadowcolor=black@0.85'
      )
      if (fontPath) draw += `:fontfile='${escapeDrawtext(fontPath)}'`
      draw += `[${nextLabel}]`
      filterComplex += `;${draw}`
      current = nextLabel
    }
    mapLabel = `[${current}]`
  }

  return run([
    '-i', inputPath,
    '-filter_complex', filterComplex,
    '-map', mapLabel,
    '-map', '0:a?',
    '-c:v', 'libx264', '-preset', 'fast', '-crf', '22',
    '-c:a', 'aac', '-b:a', '128k',
    '-pix_fmt', 'yuv420p',
    '-movflags', '+faststart',
    outputPath,
  ], 'render_repurpose_export')
}

module.exports = {
  escapeDrawtext,
  resolveFont,
  wrapLines,
  renderRepurposeExport,
}
